src/components/mc/cuda/mc_cuda.c

```diff
@@ -77,46 +78,84 @@ static ucc_status_t ucc_mc_cuda_mem_free(void *ptr)
-        cudaGetLastError();
-        return UCC_ERR_NOT_SUPPORTED;
+    CUresult                     cu_err;
+    ucc_memory_type_t            mem_type;
+    void                         *base_address;
+    size_t                       alloc_length;
+
+    if (!(mem_attr->field_mask & (UCC_MEM_ATTR_FIELD_MEM_TYPE     |
+                                  UCC_MEM_ATTR_FIELD_BASE_ADDRESS |
+                                  UCC_MEM_ATTR_FIELD_ALLOC_LENGTH))) {
+        return UCC_OK;
     }
 
+    if (ptr == 0) {
+        mem_type = UCC_MEMORY_TYPE_HOST;
+    } else {
+        st = cudaPointerGetAttributes(&attr, ptr);
+        if (st != cudaSuccess) {
+            cudaGetLastError();
+            return UCC_ERR_NOT_SUPPORTED;
+        }
 #if CUDART_VERSION >= 10000
-    switch (attr.type) {
-    case cudaMemoryTypeHost:
-        *mem_type = UCC_MEMORY_TYPE_HOST;
-        break;
-    case cudaMemoryTypeDevice:
-        *mem_type = UCC_MEMORY_TYPE_CUDA;
-        break;
-    case cudaMemoryTypeManaged:
-        *mem_type = UCC_MEMORY_TYPE_CUDA_MANAGED;
-        break;
-    default:
-        return UCC_ERR_NOT_SUPPORTED;
-    }
+        switch (attr.type) {
+        case cudaMemoryTypeHost:
+            mem_type = UCC_MEMORY_TYPE_HOST;
+            break;
+        case cudaMemoryTypeDevice:
+            mem_type = UCC_MEMORY_TYPE_CUDA;
+            break;
+        case cudaMemoryTypeManaged:
+            mem_type = UCC_MEMORY_TYPE_CUDA_MANAGED;
+            break;
+        default:
+            return UCC_ERR_NOT_SUPPORTED;
+        }
 #else
-    if (attr.memoryType == cudaMemoryTypeDevice) {
-        if (attr.isManaged) {
-            *mem_type = UCC_MEMORY_TYPE_CUDA_MANAGED;
+        if (attr.memoryType == cudaMemoryTypeDevice) {
+            if (attr.isManaged) {
+                mem_type = UCC_MEMORY_TYPE_CUDA_MANAGED;
+            } else {
+                mem_type = UCC_MEMORY_TYPE_CUDA;
+            }
+        }
+        else if (attr.memoryType == cudaMemoryTypeHost) {
+            mem_type = UCC_MEMORY_TYPE_HOST;
         } else {
-            *mem_type = UCC_MEMORY_TYPE_CUDA;
+            return UCC_ERR_NOT_SUPPORTED;
+        }
+#endif
+
+        if (mem_attr->field_mask & (UCC_MEM_ATTR_FIELD_ALLOC_LENGTH |
+                                    UCC_MEM_ATTR_FIELD_BASE_ADDRESS)) {
+            cu_err = cuMemGetAddressRange((CUdeviceptr*)&base_address,
+                    &alloc_length, (CUdeviceptr)ptr);
+            if (cu_err != CUDA_SUCCESS) {
+                mc_error(&ucc_mc_cuda.super,
+                         "ccuMemGetAddressRange(%p) error: %d(%s)",
+                          ptr, cu_err, cudaGetErrorString(st));
+                return UCC_ERR_NOT_SUPPORTED;
+            }
         }
     }
-    else if (attr.memoryType == cudaMemoryTypeHost) {
-        *mem_type = UCC_MEMORY_TYPE_HOST;
-    } else {
-        return UCC_ERR_NOT_SUPPORTED;
+
+    if (mem_attr->field_mask & UCC_MEM_ATTR_FIELD_MEM_TYPE) {
```

Comment:
Generic question (maybe i'm paranoid). Isn't that a lot of branching which is not always needed? You kept separate APIs in the top component: for mem_type and mem_query. Maybe you could keep them separate internally in the mc as well. And mem_query could call mem_type if needed. wdyt?

Reply:
I guess we already in slow path in querying attributes with CUDA api (slow ~-0.5 us) w/o mem type cache.
but,  i think I can remove these additional if () and move them to above where they computed.

Reply:
btw, do we have plans to introduce mem_type cache?

Reply:
i don't think we need it for the first release. but maybe in the future.